File: tasks/circuit-breaker/go/environment/src/breaker.go
package main

type breaker struct {
	state    string
	failures int
	openedAt int
	probes   int
}

func newBreaker() *breaker {
	return &breaker{state: "closed"}
}

// observe reports the state the breaker is in when a call arrives.
func observe(item *breaker, at int, settings config) string {
	return item.state
}

func admit(item *breaker, observed string, settings config) bool {
	return observed != "open"
}

func record(item *breaker, observed string, outcome string, at int, settings config) {
	if outcome == "success" {
		item.failures = 0
		return
	}
	item.failures++
	if item.failures >= settings.Threshold {
		item.state = "open"
		item.openedAt = at
	}
}

File: tasks/circuit-breaker/typescript/environment/src/breaker.ts
import type { Config, Outcome, State } from "./parse";

export interface Breaker {
  state: State;
  failures: number;
  openedAt: number;
  probes: number;
}

export function newBreaker(): Breaker {
  return { state: "closed", failures: 0, openedAt: 0, probes: 0 };
}

/** The state the breaker is in when a call arrives. */
export function observe(breaker: Breaker, at: number, config: Config): State {
  return breaker.state;
}

export function admit(
  breaker: Breaker,
  observed: State,
  config: Config,
): boolean {
  return observed !== "open";
}

export function record(
  breaker: Breaker,
  observed: State,
  outcome: Outcome,
  at: number,
  config: Config,
): void {
  if (outcome === "success") {
    breaker.failures = 0;
    return;
  }
  breaker.failures += 1;
  if (breaker.failures >= config.threshold) {
    breaker.state = "open";
    breaker.openedAt = at;
  }
}
